### PR TITLE
Better Spring MVC exceptions logging and handling

### PR DESCRIPTION
**Pull request by sdeleuze**

### CODE REVIEW

File: resthub-web/resthub-web-server/src/main/java/org/resthub/web/LoggingHandlerExceptionResolver.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    protected Logger logger = LoggerFactory.getLogger(LoggingHandlerExceptionResolver.class);
+    
+    @Override
+    public int getOrder() {
+        return Integer.MIN_VALUE; // we're first in line, yay!
+    }
+
+    @Override
+    public ModelAndView resolveException(HttpServletRequest request, HttpServletResponse response, Object handler, Exception e) {
+        logger.warn("Exception catched by Spring MVC : " + e);
```
Comment (loicfrering) [inline resthub-web/resthub-web-server/src/main/java/org/resthub/web/LoggingHandlerExceptionResolver.java]: There should be no space before the colon: "Exception catched by Spring MVC: "

File: resthub-web/resthub-web-server/src/main/java/org/resthub/web/ResthubHandlerExceptionResolver.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import javax.inject.Named;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.web.servlet.HandlerExceptionResolver;
+import org.springframework.web.servlet.ModelAndView;
+
+@Named("resthubHandlerExceptionResolver")
+public class ResthubHandlerExceptionResolver implements HandlerExceptionResolver {
```
Comment (loicfrering) [inline resthub-web/resthub-web-server/src/main/java/org/resthub/web/ResthubHandlerExceptionResolver.java]: We should consider extending [AbstractHandlerExceptionResolver](https://github.com/SpringSource/spring-framework/blob/master/spring-webmvc/src/main/java/org/springframework/web/servlet/handler/AbstractHandlerExceptionResolver.java) which provides common functionalities.

File: resthub-web/resthub-web-server/src/main/java/org/resthub/web/ResthubHandlerExceptionResolver.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                        handler);
+            }
+
+        } catch (Exception handlerException) {
+            logger.warn("Handling of [" + ex.getClass().getName() + "] resulted in Exception", handlerException);
+        }
+        return null;
+    }
+
+    protected ModelAndView handleIllegalArgumentHandlingMethod(IllegalArgumentException ex, HttpServletRequest request, HttpServletResponse response, Object handler) throws IOException {
```
Comment (loicfrering) [inline resthub-web/resthub-web-server/src/main/java/org/resthub/web/ResthubHandlerExceptionResolver.java]: According to conventions in Spring's DefaultHandlerExceptionResolver, method's name should be `handleIllegalArgument`.

File: resthub-web/resthub-web-server/src/main/java/org/resthub/web/ResthubHandlerExceptionResolver.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                return handleValidation((ValidationException) ex, request, response, handler);
+            } else if (ex instanceof NotFoundException) {
+                return handleNotFound((NotFoundException) ex, request, response, handler);
+            } else if (ex instanceof NotImplementedException) {
+                return handleNotImplemented((NotImplementedException) ex, request, response, handler);
+            }           
+            
+
+        } catch (Exception handlerException) {
+            logger.warn("Handling of [" + ex.getClass().getName() + "] resulted in Exception", handlerException);
```
Comment (bclozel) [inline resthub-web/resthub-web-server/src/main/java/org/resthub/web/ResthubHandlerExceptionResolver.java]: an exception while resolving an exception (INCEPTION)?
Is warn the right log level? Or should we go error?

File: resthub-web/resthub-web-server/src/main/java/org/resthub/web/controller/RepositoryBasedRestController.java [hunk 5]
```diff
@@ -70,7 +67,7 @@ public T update(@PathVariable ID id, @RequestBody T resource) {
 
         Serializable entityId = this.getIdFromResource(resource);
         if ((entityId == null) || (!id.equals(entityId))) {
-            throw new BadRequestException();
+            throw new IllegalArgumentException("No resource with id " + id + "found");
```
Comment (loicfrering) [inline resthub-web/resthub-web-server/src/main/java/org/resthub/web/controller/RepositoryBasedRestController.java]: Missing a space before `found`

File: resthub-web/resthub-web-server/src/main/java/org/resthub/web/ResthubHandlerExceptionResolver.java (hunk 4 above)
Comment (loicfrering) [inline resthub-web/resthub-web-server/src/main/java/org/resthub/web/ResthubHandlerExceptionResolver.java]: I think that @sdeleuze used the same log level as in [Spring's DafaultHandlerExceptionResolver](https://github.com/SpringSource/spring-framework/blob/master/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/support/DefaultHandlerExceptionResolver.java#L152).

Comment (sdeleuze) [inline resthub-web/resthub-web-server/src/main/java/org/resthub/web/ResthubHandlerExceptionResolver.java]: I did but you're right, error is better

File: resthub-common/src/main/java/org/resthub/common/exception/NotFoundException.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package org.resthub.common.exception;
+
+/**
+ * Exception thrown when not result where found (for example findById with null return value)
```
Comment (bclozel) [inline resthub-common/src/main/java/org/resthub/common/exception/NotFoundException.java]: Exception thrown when no result was found (for example findById with null return value)

File: resthub-common/src/main/java/org/resthub/common/exception/NotImplementedException.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package org.resthub.common.exception;
+
+/**
+ * Exception use to signify that an element (class, webservice, method) does not exists
```
Comment (bclozel) [inline resthub-common/src/main/java/org/resthub/common/exception/NotImplementedException.java]: Exception used to signify that an application element (class, webservice, method) does not exist

**Comment by bclozel** [comment]
Good to go!

**Pull request merged**